### ISSUE TITLE
Spread operator breaks observable... why?

### ISSUE THREAD

**Issue by awei01** [mobxjs/mobx#1968]
Welcome to MobX. Please provide as much relevant information as possible!

I have a:

1. [x ] Question: Feel free to just state your question. For a quick answer, there are usually people online at our [Gitter](https://gitter.im/mobxjs/mobx) channel

This code breaks mobx's reactivity.
```
const _counter = observable.box(0)
setInterval(action(() => {
  _counter.set(_counter.get() + 1)
}), 100)

const thing = {
  ...{ foo: 'bar' },
  get counter () {
    return _counter.get()
  }
}

autorun(() => {
  console.log(thing.counter)
})
```

With the following edit, it will work.
```
const _counter = observable.box(0)
setInterval(action(() => {
  _counter.set(_counter.get() + 1)
}), 100)

const thing = {
  // ...{ foo: 'bar' }, COMMENTED THIS LINE
  get counter () {
    return _counter.get()
  }
}

autorun(() => {
  console.log(thing.counter)
})
```

TLDR: spread syntax breaks mobx reactivity. I'm *guessing* that it is because of the transpilation implementation of spread syntax. But, any insight here is appreciated.

Thanks

**Comment by FredyC** [comment]
But the `thing` is not observable, how do you expect it to be one? :) Can you perhaps provide a CodeSandbox example?

**Comment by awei01** [comment]
It is observable... The getter reevaluates every second in autorun because
the underlying source is observable.

On my phone at the moment but will provide a fiddle when I get to desktop.

**Comment by FredyC** [comment]
Well, this piece alone cannot really work. The `thing.get` is not observable, it's a current value of `_counter` in the moment of call.

```js
const thing = {
  get counter () {
    return _counter.get()
  }
}
```

To make truly observable, you should wrap that object into another `observable`. Cannot imagine why would spread cause any harm there. The working example would be the best really.

**Comment by awei01** [comment]
Sorry, my note was not clear. I was trying to save keystrokes and made it confusing... I've edited original post

**Comment by awei01** [comment]
Incidentally, I cannot get it to reproduce in a fiddle yet... maybe will have to create a repo... More to come.

**Comment by FredyC** [comment]
Ok, I made reproduction and it indeed behaves the way you said. Interestingly enough, doing spread after the get works also.

https://codesandbox.io/s/priceless-montalcini-yi4c8

To be honest I am not sure why it behaves like that. Did it work in some previous version of MobX? Which one?

**Comment by awei01** [comment]
I don't know about previous versions. I just ran into this and was left
scratching my head.

**Comment by FredyC** [comment]
Ah ok, I think I get it now, it's more of the bug of compilation...

```js
var withSpreadStart = _objectSpread({}, foo, {
  get counter() {
    return _counter.get();
  }
});

var withSpreadEnd = _objectSpread({
  get counter() {
    return _counter.get();
  }
}, foo);
```

Notice that when the spread is at the beginning, it actually assign everything into a new object. Whilst in the second case it assigns the `foo` to the same object.

And the `_objectSpread` utilizes the `defineProperty` underneath which I assume breaks things here.

https://github.com/babel/babel/blob/3aaafae053fa75febb3aa45d45b6f00646e30ba4/packages/babel-helpers/src/helpers.js#L400

It's really interesting, I am using spread all the time, but haven't noticed 

---

Btw, TypeScript compiler does much better compilation [without such weird logic](https://www.typescriptlang.org/play/#src=const%20withSpreadStart%20%3D%20%7B%0D%0A%20%20...foo%2C%0D%0A%20%20get%20counter()%20%7B%0D%0A%20%20%20%20return%20_counter.get()%0D%0A%20%20%7D%0D%0A%7D%0D%0A%0D%0Aconst%20withSpreadEnd%20%3D%20%7B%0D%0A%20%20get%20counter()%20%7B%0D%0A%20%20%20%20return%20_counter.get()%0D%0A%20%20%7D%2C%0D%0A%20%20...foo%0D%0A%7D%0D%0A). Unfortunately, with Babel TypeScript support the TSC is no longer used that much.

**Comment by FredyC** [comment]
More research shows the obvious culprit. It's more of the feature/bug of language I guess. Reassigning getter to another object makes it not a getter anymore, but just a static value.

![image](https://user-images.githubusercontent.com/1096340/57970307-fcc3ab00-797f-11e9-8e1f-2744c8ebd5e7.png)

So, in the end, it's not a problem of MobX. I don't we can really do anything about it. Perhaps @mweststrate will have some idea, but closing the issue for now.

**Comment by awei01** [comment]
@FredyC Thanks for tracking it down.
